File: stacks/art_vehicle/art_nav/src/pilot/speed.py
#!/usr/bin/env python
"""
/* -*- mode: C++ -*-
 *
 *  Copyright (C) 2005, 2007, 2009 Austin Robot Technology
 *
 *  License: Modified BSD Software License Agreement
 * 
 *  $Id$
 */

/**  \file
 
     ART autonomous vehicle speed controller interface.
 
        SpeedControl -- virtual base speed controller class
   
        SpeedControlMatrix -- speed control using acceleration matrix

        SpeedControlPID -- speed control using direct PID of throttle
                           and brake

     \author Jack O'Quin, David Kraft-Ishihama

 */
"""

# This file is a combination of the speed.cc and speed.h files
# written in python. I have no idea if the code here works, but
# it should have the same behavior as the C++ version.

import roslib
roslib.load_manifest('art_nav')

import rospy
import math

from art_common import pid

EPSILON_BRAKE = 0.01
EPSILON_THROTTLE = 0.01


class SpeedControl(object) :
  def __init__(self) :
    self.brake_position_ = 0.0
    self.throttle_position_ = 0.0

  def set_brake_position(self, position) :
    self.brake_position_ = position

  def set_throttle_position(self, position) :
    self.throttle_position_ = position


class SpeedControlPID (SpeedControl) :
  def __init__(self) :
    SpeedControl.__init__(self)
    self.braking_ = True
    self.brake_pid_ = pid.Pid("brake", -0.2, -0.02, -1.6, 1.0, 0.0, 5000.0)
    self.throttle_pid_ = pid.Pid("throttle", 0.12, 0.001, 0.54, 0.4, 0.0, 5000.0)

    self.configure()
    self.reset()

  def adjust(self, speed, error, throttle_req, brake_req) :
    if self.braking_ :
      ### controlling with brake: ###
      brake_req = self.brake_pid_.Update(error, speed)
      throttle_req = 0.0
      
      # If requesting brake off, switch to throttle control.
      if (True) :
      # Must check reported brake position, too.  Otherwise there
      # will be considerable overlap, applying throttle while the
      # brake is still on.  That can cause mechanical damage to the
      # transmission.
        if ((self.brake_position_ < EPSILON_BRAKE)
            and (brake_req < EPSILON_BRAKE)) :
          brake_req = 0.0              # brake off
          self.braking_ = False             # using throttle now
          self.throttle_pid_.Clear()       # reset PID controller
        
      else :     
        # Allow more overlap, to damp the oscillations that occur when
        # switching back and forth between throttle and brake.
        if (brake_req < EPSILON_BRAKE) :
          brake_req = 0.0;                  # brake off
          self.braking_ = False             # using throttle now
          self.throttle_pid_.Clear()      # reset PID controller
    
    else :
      # controlling with throttle:
      throttle_req = self.throttle_pid_.Update(error, speed);
      brake_req = 0.0

      # If requesting throttle off, switch to brake control.

      # Since throttle responds much faster than brake, it will reach
      # idle before the brake really starts engaging.  So, not
      # checking throttle_position_ here is an option, which reduces
      # latency when slowing down.
      if (throttle_req < EPSILON_THROTTLE) :
          throttle_req = 0.0           # throttle off
          self.braking_ = True         # using brake now
          self.brake_pid_.Clear()         # reset PID controller   

    return (throttle_req, brake_req)

  def configure(self) :
    self.brake_pid_.Configure()
    self.throttle_pid_.Configure()
    return

  def reset(self) :
    self.brake_pid_.Clear()
    self.throttle_pid_.Clear() 


# TODO: Finish working on this SpeedControl subclass
class SpeedControlRL (SpeedControl) :
  def __init__ (self, training=True) :
    SpeedControl.__init__(self)
    self.training = training
    self.epsilon = 0.5
    self.alpha = 0.5
    self.gamma = 1.0
    self.weight = {}
    for action in self.legalActions() :
      self.weight[action] = {}    
    self.previousState = {}
    self.currentState = {}
    self.action = None

  def adjust(self, speed, error, throttle_req, brake_req) :
    # Update states
    self.previousState = {}
    for key in self.currentState :
      self.previousState[key] = self.currentState[key]
    self.updateCurrentState(speed, error)

    # Make training update
    if self.training :
      if self.action != None : self.update()

    # Choose an action
    maxAction = -1.0
    maxScore = self.getQValue(self.currentState, -1.0)
    for action in self.legalActions() :
      score = self.getQValue(self.currentState, action)
      if score > maxScore :
        maxScore = score
        maxAction = action
    self.action = maxAction

    # Translate the action into throttle and brake requests
    if self.action < 0.0 :
      throttle_req = 0.0
      brake_req = -self.action
    else :
      throttle_req = self.action
      brake_req = 0.0

    # Return throttle and brake requests
    return throttle_req, brake_req

  def updateCurrentState(self, speed, error) :
    # TODO: Get better state representation?
    x = 100
    while x > 0 :
      self.currentState['speed' + str(x)] = self.currentState.get('speed' + str(x - 1), 0.0)
      self.currentState['error' + str(x)] = self.currentState.get('error' + str(x - 1), 0.0)
      self.currentState['action' + str(x)] = self.currentState.get('action' + str(x - 1), 0.0)
      x -= 1     
    self.currentState['action0'] = self.action
    self.currentState['speed0'] = self.currentState.get('speed', 0.0)
    self.currentState['error0'] = self.currentState.get('error', 0.0)
    self.currentState['speed'] = speed
    self.currentState['error'] = error

  def update(self) :
    reward = self.reward()
    correction = (reward + self.gamma * self.getValue(self.currentState)) - self.getQValue(self.previousState, self.action)
    for key in self.currentState.keys() :
      self.weight[self.action][key] = self.weight[self.action].get(key, 0.0) + self.alpha * correction * self.currentState[key]
    return

  def reward(self) :
    # TODO: Make a better reward function
    if math.fabs(self.currentState['error']) < 1.0 :
      return 2.0
    if math.fabs(self.currentState['error']) < math.fabs(self.currentState['error0']) :
      return 1.0
    if math.fabs(self.currentState['error']) > math.fabs(self.currentState['error0']) :
      return -1.0
    return 0.0

  def legalActions(self) :
    actions = []
    temp = -1.0
    while temp <= 1.0 :
      actions += [temp]
      temp += 0.25
    return actions

  def getQValue(self, state, action) :
    sum = 0.0
    for key in state.keys() :
      sum += state.get(key) * self.weight[action].get(key, 0.0)
    return sum
    
  def getValue(self, state) :
    return max([self.getQValue(state, action) for action in self.legalActions()])

  def configure(self) :
    #self.update()
    return

  def reset(self) :
    # What to do with reset? Nothing?
    return

File: stacks/art_vehicle/art_pilot/src/pilot/speed.cc
/*
 *  Copyright (C) 2005, 2007, 2009, 2011 Austin Robot Technology
 *  License: Modified BSD Software License Agreement
 * 
 *  $Id$
 */

/**  @file
 
     ART autonomous vehicle speed controller.

     @author Jack O'Quin
 */

#include <art/conversions.h>
#include <art_msgs/ArtHertz.h>
#include "speed.h"

/**
 @brief ART autonomous vehicle speed controller.
*/

namespace ControlMatrix
{

  // acceleration matrix dimensions
  static const int N_DELTAS = 13;
  static const int DELTA_0 = (N_DELTAS-1)/2; // middle row: delta == 0
  static const int N_SPEEDS = 6;

  // Velocity control
  typedef struct
  {
    float       brake_delta;
    float       throttle_delta;
  } accel_parms_t;

  // Acceleration matrix: rows are indexed according to the requested
  // speed delta in MPH, columns are indexed by current speed in MPH.
  // The main reason for making all this table-driven is so we can
  // adjust some values without affecting all the others.  
  //
  // NOTE: values in this table represent a percentage change per second,
  // so their effect builds up quickly.

  static const accel_parms_t accel_matrix[N_DELTAS][N_SPEEDS] =
    { /*              0        <=4       <=8       <=16      <=32      more  */
      /* -32 */ {{1000,-90}, {12,-30}, {12,-30}, {12,-40}, {12,-50}, {20,-70}},
      /* -16 */ {{1000,-60}, { 6,-20}, { 6,-20}, { 6,-30}, { 6,-40}, {10,-55}},
      /*  -8 */ {{1000,-40}, { 4,-15}, { 4,-15}, { 4,-20}, { 4,-30}, { 7,-40}},
      /*  -4 */ { {500,-20}, { 3,-10}, { 3,-10}, { 3,-10}, { 3,-20}, { 5,-25}},
      /*  -2 */ { {200,-10},  {2,-5},   {2,-5},   {2,-5},   {2,-10},  {3,-10}},
      /*  -1 */ { {100,-5},   {1,-2},   {1,-2},   {1,-2},   {1,-5},   {1,-5} },
      /*   0 */ { {100,0},    {0,0},    {0,0},    {0,0},    {0,0},    {0,0}  },
      /*   1 */ {{-100,1},    {-2,2},   {-2,2},   {-2,2},   {-2,5},   {-2,5} },
      /*   2 */ {{-200,2},    {-5,5},   {-5,5},   {-5,5},   {-5,10},  {-5,10}},
      /*   4 */ {{-300,3},   {-10,10}, {-10,10}, {-10,10}, {-10,20}, {-10,25}},
      /*   8 */ {{-500,4},   {-15,15}, {-15,15}, {-15,20}, {-15,30}, {-15,40}},
      /*  16 */ {{-1000,5},  {-20,20}, {-20,20}, {-20,30}, {-20,40}, {-20,55}},
      /*  32 */ {{-1000,6},  {-30,30}, {-30,30}, {-30,40}, {-30,50}, {-30,70}}
    };

  // If it turns out that the matrix indices should be different, adjust
  // the conversions in these functions.  The logarithmic intervals are
  // based on intuition.
  static inline int
  delta_row(float mph)
  {
    int row;
    static float row_limits[DELTA_0+1] = {0.0, 1.0, 2.0, 4.0, 8.0, 16.0, 32.0};
    int mph_sign = 1;

    if (mph < 0)
      mph_sign = -1;

    float abs_mph = mph * mph_sign;

    for (row = 0; row < DELTA_0+1; ++row)
      {
        if (abs_mph <= row_limits[row])
          return DELTA_0 + mph_sign * row;
      }

    // if delta outside the table, use the first or last entry
    return DELTA_0 + mph_sign * DELTA_0;
  }

  static inline int
  speed_col(float mph)
  {
    int col;
    static float col_limits[N_SPEEDS] = {0.0, 4.0, 8.0, 16.0, 32.0, 64.0};

    for (col = 0; col < N_SPEEDS; ++col)
      {
        if (mph <= col_limits[col])
          return col;
      }

    // if speed beyond the table, use the last entry
    return N_SPEEDS-1;
  }
};

/** Acceleration matrix speed control constructor. */
SpeedControlMatrix::SpeedControlMatrix():
  SpeedControl(),
  velpid_(new Pid("speed", 2.0, 0.0, 32.0))
{
  // Allocate speed PID class with default parameters, then
  // configure control constants.
  //
  // Default PD parameters: make Kd proportional to cycle rate, higher
  // frequencies will yield lower deltaV per cycle.
  //
  reset();
}

/** SpeedControlMatrix destructor */
SpeedControlMatrix::~SpeedControlMatrix()
{}

/** Adjust speed to match goal.

    Generate brake and throttle changes from velocity PID controller
    via an acceleration matrix.

    @param speed absolute value of current velocity in m/sec
    @param error immediate goal minus speed
    @param brake_req -> previous brake request (input),
                        updated brake request (output).
    @param throttle_req -> previous throttle request (input),
                        updated throttle request (output).
*/
void SpeedControlMatrix::adjust(float speed, float error,
                                float *brake_req, float *throttle_req)
{
  using namespace ControlMatrix;

  float delta = velpid_->Update(error, speed);
  float delta_mph = mps2mph(delta);

  // index into the acceleration matrix
  int row = delta_row(delta_mph);
  int col = speed_col(mps2mph(speed));
  float brake_delta =
    (accel_matrix[row][col].brake_delta / art_msgs::ArtHertz::PILOT)
    / 100.0;
  float throttle_delta =
    (accel_matrix[row][col].throttle_delta / art_msgs::ArtHertz::PILOT)
    / 100.0;

  ROS_DEBUG("accel_matrix[%d][%d] contains {%.3f, %.3f}",
            row, col, brake_delta, throttle_delta);

  // Do not add braking unless nearly idle throttle was previously
  // requested, or throttle unless the brake is nearly off.
  if (*throttle_req > 0.0 && brake_delta > 0.0)
    brake_delta = 0.0;
  if (*brake_req > 0.0 && throttle_delta > 0.0)
    throttle_delta = 0.0;
  *brake_req += brake_delta;
  *throttle_req += throttle_delta;
}

/** Configure controller parameters. */
void SpeedControlMatrix::configure(art_pilot::PilotConfig &newconfig)
{
  velpid_->Configure(node_);
}

/** Reset speed controller. */
void SpeedControlMatrix::reset(void)
{
  velpid_->Clear();
}

/** PID speed control constructor.
 *
 *  Allocate brake and throttle PID classes with default parameters,
 *  then configure control constants.
 *
 *  Throttle gains came from Jesse Tannahill's pilot prototype.
 *  Brake gains selected by experiment.
 */
SpeedControlPID::SpeedControlPID():
  SpeedControl(),
  braking_(true),                       // begin with brake on
  brake_pid_(new Pid("brake", -0.2, -0.0002, -1.6, 1.0, 0.0, 5000.0)),
  throttle_pid_(new Pid("throttle", 0.12, 0.001, 0.54, 0.4, 0.0, 5000.0))
{
  reset();
}

/** PID speed control destructor. */
SpeedControlPID::~SpeedControlPID()
{};

/** Adjust speed to match goal.

    Generate brake and throttle changes directly from separate PID
    controllers.  Uses a state machine to avoid applying simultaneous
    brake and throttle.

    @param speed absolute value of current velocity in m/sec
    @param error immediate goal minus speed
    @param brake_req -> previous brake request (input),
                        updated brake request (output).
    @param throttle_req -> previous throttle request (input),
                        updated throttle request (output).
*/
void SpeedControlPID::adjust(float speed, float error,
                             float *brake_req, float *throttle_req)
{
  if (braking_)
    {
      // controlling with brake:
      *brake_req = brake_pid_->Update(error, speed);
      *throttle_req = 0.0;
      
      // If requesting brake off, switch to throttle control.
#if 1
      // Must check reported brake position, too.  Otherwise there
      // will be considerable overlap, applying throttle while the
      // brake is still on.  That can cause mechanical damage to the
      // transmission.
      if ((brake_position_ < EPSILON_BRAKE) && (*brake_req < EPSILON_BRAKE))
        {
          *brake_req = 0.0;             // brake off
          braking_ = false;             // using throttle now
          throttle_pid_->Clear();       // reset PID controller
        }
#else
      // Allow more overlap, to damp the oscillations that occur when
      // switching back and forth between throttle and brake.
      if ((*brake_req < EPSILON_BRAKE))
        {
          *brake_req = 0.0;             // brake off
          braking_ = false;             // using throttle now
          throttle_pid_->Clear();       // reset PID controller
        }
#endif
    }
  else
    {
      // controlling with throttle:
      *throttle_req = throttle_pid_->Update(error, speed);
      *brake_req = 0.0;

      // If requesting throttle off, switch to brake control.

      // Since throttle responds much faster than brake, it will reach
      // idle before the brake really starts engaging.  So, not
      // checking throttle_position_ here is an option, which reduces
      // latency when slowing down.
      if (*throttle_req < EPSILON_THROTTLE)
        {
          *throttle_req = 0.0;          // throttle off
          braking_ = true;              // using brake now
          brake_pid_->Clear();          // reset PID controller
        }
    }
}

/** Configure controller parameters. */
void SpeedControlPID::configure(art_pilot::PilotConfig &newconfig)
{
  brake_pid_->Configure(newconfig.brake_kp,
                        newconfig.brake_ki,
                        newconfig.brake_kd);
  throttle_pid_->Configure(newconfig.throttle_kp,
                           newconfig.throttle_ki,
                           newconfig.throttle_kd);
}

/** Reset speed controller. */
void SpeedControlPID::reset(void)
{
  brake_pid_->Clear();
  throttle_pid_->Clear();
}
